spec/justified.tla

```diff
@@ -0,0 +1,209 @@
+---- MODULE justified ----
+
+EXTENDS ffg, Sequences
+
+\* Given a set of votes, returns all of the checkpoints, which appear as sources.
+\* @type: (Set($signedVoteMessage)) => Set($checkpoint);
+Sources(votes) == { vote.message.ffg_source : vote \in votes} 
+
+\* The recursion in is_justified_checkpoint is set up as follows:
+\* To justify checkpoint C, we must look at all votes C1 -> C2, s.t.
+\* C lies between C1 and C2 (definition below), AND C1 is justified.
+\* We define the following operator, which precisely captures this condition, 
+\* save for the recursive justificaiton.
+\* @type: ($signedVoteMessage, $checkpoint, $commonNodeState) => Bool;
+IsVoteInSupportAssumingJustifiedSource(vote, checkpoint, node_state) ==
+    \* TODO: If we end up always doing 1-state model checking, a lot of sub-checks in 
+    \* valid_FFG_vote are already a part of the state validity predicate, and couldbe omitted here 
+    \* to simplify computation
+    /\ valid_FFG_vote(vote, node_state) 
+    /\ vote.message.ffg_target.chkp_slot = checkpoint.chkp_slot
+    \* TODO: If we end up always doing 1-state model checking, we could 
+    \* precompute the relation is_ancestor_descendant_relationship(a,b), for all blocks
+    \* ahead of time as a _function_, s.t. this lookup here becomes a simple access, instead of each of them being
+    \* another pseudo-recursive computation
+    /\ is_ancestor_descendant_relationship(
+        get_block_from_hash(checkpoint.block_hash, node_state),
+        get_block_from_hash(vote.message.ffg_target.block_hash, node_state),
+        node_state
+        )
+    /\ is_ancestor_descendant_relationship(
+        get_block_from_hash(vote.message.ffg_source.block_hash, node_state),
+        get_block_from_hash(checkpoint.block_hash, node_state),
+        node_state
+        )
+
+\* With the predicate, we can filter out the votes that are relevant in one step
+\* @type: ($checkpoint, $commonNodeState) => Set($signedVoteMessage);
+VotesRelevantForOneStepIteration(checkpoint, node_state) == 
```

Comment:
Since the predicate is called `IsVoteInSupportAssumingJustifiedSource`, it might make sense to call this

```suggestion
VotesInSupportAssumingJustifiedSource(checkpoint, node_state) == 
```